### ISSUE TITLE
Start with "dev mode" on when booting into the console

### ISSUE THREAD

**Issue by blinry** [nesbox/TIC-80#1950]
**Problem**

When opening https://tic80.com/create or running any of the official releases, "dev mode" is off in the settings. This means that after running a game, upon hitting Esc, people have to select "close game" from the menu every time they want to return to the console. Also, the shortcuts F1 to F5 don't work to return to specific editors.

This greatly hinders the development flow, especially for new people who don't know that they can disable this behavior in the settings. I think that both advanced programmers as well as newcomers would much prefer having dev mode on when they want to develop a game. Also personally, disabling dev mode would be the first thing I'd always do when opening TIC-80. So I think this should be the default behavior.

When I gave a workshop on TIC-80 last month, it was unfortunate to have to explain how to enable dev mode to each of the over 20 participants. I would rather want a smooth first-time experience for them, where they can switch between making and trying the game quickly, without having to find the correct menu entry before.

**Proposed solution**

When booting the console without immediately starting a game, enable dev mode. In this case, it is very likely that this is the desired behavior.

@nesbox [pointed out](https://github.com/nesbox/TIC-80/discussions/1894#discussioncomment-2889705) that the menu would also be good to have when using `surf`. I'd propose to disable dev mode again when running `surf`, in that case.

**Comment by borbware** [comment]
Big agree!

Also, while we're at it, make ESC go to console in devmode instead of code editor, like it used to be.

These minor changes in workflow have basically made me stay in 0.90 for the time being...

**Comment by darltrash** [comment]
In my opinion, A thing I like a lot about TIC-80 is how much it involves the average user into making things and dividing developers from regular users is not helpful in general.
I think it'd be better if we could just jump back into the editor with a shortcut of some sorts, kind of like if we could use CTRL+R to go back to the editor and NOT disable the menu or whatever at all.

**Comment by borbware** [comment]
> I simply don't think the dev mode is either useful or necessary

Dev mode is necessary! For so many TIC-80 users, "press ESC to go to console" is deeply ingrained to their muscle memory, so I can't be the only one to want that functionality back.

**Comment by borbware** [comment]
> At least make dev mode easy to toggle
imo, it IS easy to toggle. `menu` -> options > dev mode

**Comment by blinry** [comment]
Maybe as a clarifying question to @nesbox: why does the menu (and the "non-dev mode") exist at all? I always imagined that it served two purposes:

1. Make it easy for "players" to modify the key layout (by making that setting visible in the menu that opens when you hit escape). I think it serves this purpose pretty well.
2. Shield the "players" from the internals of TIC-80, so that they are not confused when they suddenly see white text on black, or strange editors. I agree with @darltrash that I would prefer it if TIC-80 would nudge the "players" a bit, by giving them an easy peek like "see, this is how the game is made! You can make games, too!" <3

But maybe there are more purposes for the menu! :)

**Comment by darltrash** [comment]
> Maybe as a clarifying question to @nesbox: why does the menu (and the "non-dev mode") exist at all? I always imagined that it served two purposes:
> 
> 
> 
> 1. Make it easy for "players" to modify the key layout (by making that setting visible in the menu that opens when you hit escape). I think it serves this purpose pretty well.
> 
> 2. Shield the "players" from the internals of TIC-80, so that they are not confused when they suddenly see white text on black, or strange editors. I agree with @darltrash that I would prefer it if TIC-80 would nudge the "players" a bit, by giving them an easy peek like "see, this is how the game is made! You can make games, too!" <3
> 
> 
> 
> But maybe there are more purposes for the menu! :)

No, no, no, I am not against the menu, I think the menu is a great idea and it's very nice to use in general, I am against the menu always showing up and getting in the way when I just want to test out the code changes I did, Maybe the default could be that there's no menu when you run the game from an editor and use a shortcut to make the menu display ...maybe

**Comment by darltrash** [comment]
> > At least make dev mode easy to toggle
> 
> 
> 
> imo, it IS easy to toggle. `menu` -> options > dev mode 

oh that does make sense, yeah

**Comment by TimotheeGreg** [comment]
I have no opinion on this question, but note that when you are on the menu, the F1 to F5 shortcuts work. So instead of typing Esc like v0.9 you can type Esc + F1 which is nearly as easy as they are adjacent keys. I never used dev mode thanks to this shortcut.
However, if people don't know that, it remains a problem. Suggesting in some way to type F1 in the menu could be a solution ?

**Comment by potato25** [comment]
I strongly agree with this. The cart should start with "dev mode" on. I'm tired of pressing 'Esc' and choosing 'dev mode on' every time

**Comment by aliceisjustplaying** [comment]
I want this too and I started working on this but turns out it's more tricky than I thought, so we'll see.

**Comment by Skeptim** [comment]
Could this be a solution: https://github.com/nesbox/TIC-80/issues/2531 ?
If we could keep the settings when updating TIC-80 one would have to choose "dev mode on" only once and for all.

**Comment by aliceisjustplaying** [comment]
Honestly, yeah. I'll check how much work it is to remove the per-release new settings feature...

**Comment by potato25** [comment]
I'm tired of pressing 'Esc' and choosing 'dev mode on' every time for all carts on the website. Plz give account setting that would let users set their dev mode and it's applied for all carts for that user. Since this is independent of TIC dev ver 1.2 etc, i think this could be implemented in the website directly now

**Comment by hw762** [comment]
If I didn't see this issue, I would have never knew dev mode existed.

The game menu made it look like it's for "release mode configuration", so I had no idea that option was there. I always used F1-F5

**Comment by Skeptim** [comment]
> The game menu made it look like it's for "release mode configuration"

Indeed the description is currently "The game menu is disabled in dev mode" which is not so clear.
What if we write "Disable menu to close game with ESC"? Or "Disable menu to switch to editor with ESC"?
Other suggestions?

> I had no idea that option was there. I always used F1-F5

Was it a pain to use F1-F5?
I wonder what is the impact of habits here.

**Comment by potato25** [comment]
My 'Esc' key broken by pressing 'Esc' so many times on the website. It's a laptop and i can't change keyboard. I would prefer default to be 'dev mode on'. I would prefer into going into code editor by pressing F1,F2, etc. But i highly doubt that these changes would be pushed out soon :/

**Comment by Skeptim** [comment]
> Was it a pain to use F1-F5?

I mean using Esc+F1 - Esc+F5

I think nesbox does not want dev mode to be 'on' by default because that's necessary for the games that use the [menu API](https://github.com/nesbox/TIC-80/wiki/MENU) and I agree with that.
I also prefer having the menu easily accessible for newcomers… if and only if they can learn to use Esc+F1 - Esc+F5, but I made a PR to write "Press F1 to switch to editor" in the menu https://github.com/nesbox/TIC-80/pull/2570 so that should be ok.

> My 'Esc' key broken by pressing 'Esc' so many times on the website. It's a laptop and i can't change keyboard. I would prefer default to be 'dev mode on'. I would prefer into going into code editor by pressing F1,F2, etc. But i highly doubt that these changes would be pushed out soon :/ It also feels like a punishment for game devs who just want to see the code. Why go through the long process of 'Esc' -> ' Options' -> 'Dev mode on' -> ' Resume game'

Your situation is unique and I understand how annoying it is, but when 'Esc' works game devs can just do 'Esc+F1' when dev mode is off, which does not seem to be a punishment to me.

**Comment by potato25** [comment]
Alternative is to set dev mode option using meta tag

**Comment by Skeptim** [comment]
Oh yeah! That solves everything, that's the best solution! Thanks!

**Comment by potato25** [comment]
> Alternative is to set dev mode option using meta tag. In cart, dev mode tag is 'on' by default. Those who want to see menu can set the meta tag 'off'. @nesbox agrees with this and asks to assign this issue to him

Plz assign this issue to nesbox. I don't know how to assign it to him

**Comment by Skeptim** [comment]
I don't think I can, or I don't know how.

**Comment by aliceisjustplaying** [comment]
> @nesbox agrees with this and asks to assign this issue to him

where?

**Comment by potato25** [comment]
> > @nesbox agrees with this and asks to assign this issue to him
> 
> where?

Here

![Screenshot 2024-05-08 144243](https://github.com/nesbox/TIC-80/assets/56357462/ea817df1-d630-4eb7-9bf7-c1f7eceb7fec)

**Comment by aliceisjustplaying** [comment]
@potato25 thanks. assigned it to @nesbox.

**Comment by hw762** [comment]
ESC+F1-F5 was OK.

In fact, my new problem is that I can't find out how to turn dev mode off

**Comment by Skeptim** [comment]
@hw762 That's a recurrent issue when we discover dev mode ^^ Use `menu` in the console

**Comment by potato25** [comment]
> In fact, my new problem is that I can't find out how to turn dev mode off

Meta tag solves this

**Comment by nesbox** [comment]
@potato25 here is the PR #2581 with `devmode: on` metatag

**Comment by blinry** [comment]
Thanks for working on this, @nesbox! :) This is an interesting addition, but it doesn't solve my issue. Let me explain why!

I think what we need is two things at the same time (and please, someone correct me if my assumptions here are wrong):

1. :technologist: Developers should have a good experience developing the game, with fast iteration cycles. This means being able to have a single-button way to switch between coding and trying the game. Esc seems like a good button for this, in my opinion.
2. :video_game: Players should be protected from leaving the game when they don't want to. This means that Esc shouldn't stop the game for them.

Right now, (2) is fulfilled by default. #2581 adds an easier way to enable (1), but it's not there by default, and it needs documentation for people to discover it. Otherwise, developers (especially new ones) need to know/be taught that they either need to go into the menu and enable dev mode, or add the meta tag.

My envisioned solution is that, if TIC-80 boots into a game directly (like in an exported game, or maybe even after going through `surf`), we adjust for the need of the players, disable dev mode, and allow (1). On the other hand, if TIC-80 boots into the console, we assume that we should adjust for the need of developers, and enable dev mode.

Two questions, @nesbox:

- Do you think this is technically feasible? Or is it maybe a problem to even tell these situations apart?
- Would you object to me drafting such a solution, and opening a PR, to see whether this is a good solution? Would be happy to work on this, I have a bit of experience with the source code from #1840.

Also curious what other folks think about this!

**Comment by potato25** [comment]
> @potato25 here is the PR #2581 with `devmode: on` metatag By default, **devmode** is disabled and we will see the game menu, but if you enable it in settings or add the `dev mode: on` meta tag, the game menu will not be shown.

Plz make 'devmode: on' as default. The whole point of this is that we don't want to see 'menu' when loaded. Those who want to see the menu can set 'devmode: off' in their carts.

**Comment by Skeptim** [comment]
> Plz make 'devmode: on' as default. The whole point of this is that we don't want to see 'menu' when loaded. Those who want to see the menu can set 'devmode: off' in their carts.

With `devmode: on` as default, past games that use the game menu won't work any more. Except if we do what @blinry suggests or if we set 'devmode: off' when 'menu' metatag is used.

> I think what we need is two things at the same time (and please, someone correct me if my assumptions here are wrong):
>
>    🧑‍💻 Developers should have a good experience developing the game, with fast iteration cycles. This means being able to have a single-button way to switch between coding and trying the game. Esc seems like a good button for this, in my opinion.

I just want to underline that in my point of view 'ESC+F1' is (nearly) as simple as 'ESC'! I think it's mostly a question of habits. Note that for switching from editor to game nobody complains about 'CTRL+R', in comparison 'ESC+F1' is easier.
Especially now that `F1` is advertised in the menu https://github.com/nesbox/TIC-80/issues/2274

However, I am not against @blinry solution. The only disadvantage would be that newcommers will need to look at the documentation to find the menu. When you use a keyboard which is not qwerty, finding the menu to remap the keys is important, for example.

**Comment by potato25** [comment]
> With `devmode: on` as default, past games that use the game menu won't work any more. Except if we do what @blinry suggests or if we set 'devmode: off' when 'menu' metatag is used.

Setting 'dev mode: on' meta tag as default probably couldn't have bigger impact than deprecating 'SCN' and 'OVR'; which nesbox did in the past

**Comment by Skeptim** [comment]
I think 'SCN' and 'OVR' still work for backward compatibility, they are just not recommended any more. Am I wrong?

**Comment by potato25** [comment]
> I think 'SCN' and 'OVR' still work for backward compatibility, they are just not recommended any more. Am I wrong?

I actually haven't used vbank() after deprecating 'OVR'. I don't understand how 'vbank' works

**Comment by potato25** [comment]
> I just want to underline that in my point of view 'ESC+F1' is (nearly) as simple as 'ESC'! I think it's mostly a question of habits if some people don't like 'ESC+F1'. Note that for switching from editor to game nobody complains about 'CTRL+R', in comparison 'ESC+F1' is easier. Especially now that `F1` is advertised in the menu #2274
 
I like 'Esc' for terminal, 'F1' for code editor, 'F2' for sprite editor etc :)

**Comment by nesbox** [comment]
@Skeptim @potato25 yes `SCN` and `OVR` still work as before, the only thing these functions don't highlight in the code editor, so I don't see problem here.

@blinry It's becoming increasingly difficult to please all parties :) you're right, players need to see the game menu, and developers need to see the code. Perhaps we really need to remove the developer option from everywhere (from the game menu and metatag) and only show the game menu in the following cases:
- the game was run from surf;
- the game was automatically run in binary for export;
- the game was run from the command line, like `tic80.exe cart.tic`;

so, if you have the desire and time let's try to do this.

**Comment by Skeptim** [comment]
> @Skeptim @potato25 yes `SCN` and `OVR` still work as before, the only thing these functions don't highlight in the code editor, so I don't see problem here.

Note that `SCN` still highlight in the code editor. That's linked to this question: https://github.com/nesbox/TIC-80/issues/2501

**Comment by potato25** [comment]
One of the solutions could be pressing unused keys liked 'F10' shows 'menu' while 'Esc' brings command line, F1 = code editor , F2= sprite editor,.... Don't know if F10 is used or not

**Comment by Skeptim** [comment]
Note that the hotkeys topic of help command could have to be modified if something new is implemented here https://github.com/nesbox/TIC-80/pull/2594

**Comment by nesbox** [comment]
So, in the end, I **removed** devmode and game menu by default, game menu will be shown if the `-- menu:` tag exists in the tags section and in case the game is exported without editors.

**Comment by imsys** [comment]
Oh, I have to open back this discussion.

I have been using TIC mostly as a player, but mainly using the libretro core.

I used the TIC PC standalone to test some mouse games, and I was using an older build.

I got the instinct to pause the game using the ESC key, the problem is that in the lastes version it brings up the code, and I can't go back to the game. There is the option to go to the console, but if I type `run`, the game starts over and I lose progress.

Now that I was typing this message, I found a way to go back to the game, but it's very inconvenient. I have to go to the terminal, type `menu` and then click `resume game`.

And I find the requirement to the games o have the `-- menu:` tag quite problematic as previous games did not have that.

If I go to my games folder that contains 1228 games and type `grep " menu:" *`, I find out that only 27 games uses that tag. (2% of the games)
Those are:

55BUTTERFEXE - 3636
5BUTTERFEXE - 3332
BABY SIMULATOR - 2586
BALMUNG - 1529
BALMUNG - 636
BITSY - 3487
BUNNYCAKE - 1440
CELESTE - 3357
COLLECT4 SOLITAIRE - 3222
COOL AND NEW RPG - 464
EMUUROM - 1378
ERAZOR - 4100
HADAL RUSH TIC - 3548
IMMUNIA - 3635
NUMBER MUNCHERS - 3257
PADDIX - 4075
POCKET HANG ON - 3786
PONG - 3883
PONG - 4176
SHOOTERTIC - 3576
SONIC MINI - 3460
SPACESHOOTERTIC - 3239
SRBKART - 4308
TICDRAW - 3896
TIC GOLF - 3578
TIC SOCCER - 4249
UNKNOWN GROUNDS - 171


So, while I was doing a fix in the android TIC80 standalone (#2837), I found that pressing the android back button is the same as pressing the ESC key, so in 98% of the cases it brings up the code and it is quite annoying.

I got some thoughts:
- Include the PAUSE key that currently is not mapped. I think this could be mapped to always show the menu.
- map android back key to the PAUSE key, so it opens the menu.

But the ideal would probably be a config file where each user could map their keys, as each person will have their own preferences.
